### PR TITLE
Detect outdated launcher

### PR DESCRIPTION
**Pull request by JaroslavTulach**
### Pull Request Description

When any of the JARs is older than the main executable, print a warning and switch to JVM mode.

### Important Notes

The `base` is obtained by `getCodeSource().getLocation()`. In case of JVM mode, it is a JAR - e.g. not executable file. In case of _native image_ launcher, it is the executable file - e.g. `base.canExecute()` returns true. As such the check is only performed in _native image_ launcher.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] All code follows the
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
- [x] Manually verified to work

### CODE REVIEW

**Review by hubertp — APPROVED** [review]

**Comment by JaroslavTulach** [comment]
Looks like the CI is affected by this change:
- [arrow language JAR](https://github.com/enso-org/enso/actions/runs/10298609673/job/28504113047?pr=10779#step:7:1287)
- [runtime-language-arrow-2024.4.1-dev.jar](https://github.com/enso-org/enso/actions/runs/10298609673/job/28504113047?pr=10779#step:7:3644)

Thanks to that two tests are running in JVM mode instead of NI mode...

File: engine/runner/src/main/java/org/enso/runner/Main.java [hunk 1]
```diff
@@ -1267,13 +1272,13 @@ private void launch(String[] args) throws IOException, InterruptedException, URI
             commandAndArgs.add(op);
           }
         }
-
+        var assertsOn = false;
+        assert assertsOn = true;
+        if (assertsOn) {
```
Comment (JaroslavTulach) [inline engine/runner/src/main/java/org/enso/runner/Main.java:1328]: This _"transfer of assertion mode"_ is now causing [following failure](https://github.com/enso-org/enso/actions/runs/10347694029/job/28641119556?pr=10779), @Akirathan:
```
Execution finished with an error: java.lang.AssertionError
         at <java> org.enso.runtime/org.enso.interpreter.runtime.warning.WithWarnings.<init>(WithWarnings.java:82)
         at <java> org.enso.runtime/org.enso.interpreter.runtime.warning.AppendWarningNode.doObjectMultipleWarningsHashMap(AppendWarningNode.java:120)
         at <java> org.enso.runtime/org.enso.interpreter.runtime.warning.AppendWarningNodeGen.executeAndSpecialize(AppendWarningNodeGen.java:318)
         at <java> org.enso.runtime/org.enso.interpreter.runtime.warning.AppendWarningNodeGen.executeAppend(AppendWarningNodeGen.java:193)
         at <java> org.enso.runtime/org.enso.interpreter.runtime.data.atom.InstantiateNode.doExecute(InstantiateNode.java:102)
         at <java> org.enso.runtime/org.enso.interpreter.runtime.data.atom.InstantiateNodeGen.executeAndSpecialize(InstantiateNodeGen.java:95)
         at <java> org.enso.runtime/org.enso.interpreter.runtime.data.atom.InstantiateNodeGen.executeGeneric(InstantiateNodeGen.java:76)
         at <java> org.enso.runtime/org.enso.interpreter.node.callable.function.BlockNode.executeGeneric(BlockNode.java:54)
         at <java> org.enso.runtime/org.enso.interpreter.node.ClosureRootNode.execute(ClosureRootNode.java:85)
         at <enso> Data.Value(Internal)
         at <enso> Data.type.new(src/Decimal_Benchmark.enso:30:9-180)
         at <enso> Decimal_Benchmark.collect_benches.Decimal_Benchmark.collect_benches<arg-1>(src/Decimal_Benchmark.enso:37:29-42)
```
the 5575d8815a change should fix the problem.

File: engine/runner/src/main/java/org/enso/runner/Main.java [hunk 2]
```diff
@@ -1230,14 +1230,19 @@ private void launch(String[] args) throws IOException, InterruptedException, URI
     var logMasking = new boolean[1];
     var logLevel = setupLogging(line, logMasking);
 
+    var loc = Main.class.getProtectionDomain().getCodeSource().getLocation();
+    var component = new File(loc.toURI().resolve("..")).getAbsoluteFile();
+    if (!component.getName().equals("component")) {
+      component = new File(component, "component");
+    }
+    assert isLauncherOutdated(new File(loc.toURI()), component) || true;
```
Comment (Akirathan) [inline engine/runner/src/main/java/org/enso/runner/Main.java]: Shouldn't this be `assert !isLauncherOutdated(...)`? The current assert never fails and only prints a warning?

Comment (JaroslavTulach) [reply engine/runner/src/main/java/org/enso/runner/Main.java]: The _"outdated launcher situation"_ can only happen and be fixed during development. If it happens in production, we have bigger problems. Development mode is currently detected by running with `-ea`. As such I am hiding this check behind `assert`. But it shouldn't fail, thus it has `|| true` at the end.

Comment (JaroslavTulach) [reply engine/runner/src/main/java/org/enso/runner/Main.java]: Condition reverted to `assert checkOutdatedLauncher` in 9014c9a, but it still remains just a printed warning.

File: engine/runner/src/main/java/org/enso/runner/Main.java (hunk 1 above)
Comment (JaroslavTulach) [reply engine/runner/src/main/java/org/enso/runner/Main.java:1328]: Alas, that fix leads to [this error](https://github.com/enso-org/enso/actions/runs/10383327368/job/28748030705#step:7:3843):
```
️ Execution finished with an error: java.lang.AssertionError
 ️at <java> org.enso.interpreter.runtime.warning.WithWarnings.<init>(WithWarnings.java:84)
 ️at <java> org.enso.interpreter.runtime.warning.AppendWarningNode.doWithWarnMultipleWarningsHashMap(AppendWarningNode.java:140)
  at <java> org.enso.interpreter.runtime.warning.AppendWarningNodeGen.executeAndSpecialize(AppendWarningNodeGen.java:345)
  at <java> org.enso.interpreter.runtime.warning.AppendWarningNodeGen.executeAppend(AppendWarningNodeGen.java:193)
  at <java> org.enso.interpreter.node.callable.InvokeMethodNode.doWarning(InvokeMethodNode.java:479)
 ️at <java> org.enso.interpreter.node.callable.InvokeMethodNodeGen.executeAndSpecialize(InvokeMethodNodeGen.java:781)
 ️at <java> org.enso.interpreter.node.callable.InvokeMethodNodeGen.execute(InvokeMethodNodeGen.java:533)
 ```
let's assume 0b9a49b fixes both - the instantiate and the invoke case.

**Review by Akirathan — APPROVED** [review]

**Pull request merged**